src/db.c

```diff
@@ -96,53 +113,17 @@ robj *lookupKey(redisDb *db, robj *key, int flags) {
  *
  *  LOOKUP_NONE (or zero): no special flags are passed.
  *  LOOKUP_NOTOUCH: don't alter the last access time of the key.
+ *  LOOKUP_NONOTIFY: don't trigger keyspace event on key miss.
+ *  LOOKUP_NOSTATS: don't increment key hits/misses couters.
+ *  LOOKUP_NOEXPIRE: don't check if the key has expired.
  *
  * Note: this function also returns NULL if the key is logically expired
  * but still existing, in case this is a slave, since this API is called only
  * for read operations. Even if the key expiry is master-driven, we can
  * correctly report a key is expired on slaves even if the master is lagging
  * expiring our key via DELs in the replication link. */
 robj *lookupKeyReadWithFlags(redisDb *db, robj *key, int flags) {
-    robj *val;
-
-    if (expireIfNeeded(db,key) == 1) {
-        /* If we are in the context of a master, expireIfNeeded() returns 1
-         * when the key is no longer valid, so we can return NULL ASAP. */
-        if (server.masterhost == NULL)
-            goto keymiss;
-
-        /* However if we are in the context of a slave, expireIfNeeded() will
-         * not really try to expire the key, it only returns information
-         * about the "logical" status of the key: key expiring is up to the
-         * master in order to have a consistent view of master's data set.
-         *
-         * However, if the command caller is not the master, and as additional
-         * safety measure, the command invoked is a read-only command, we can
-         * safely return NULL here, and provide a more consistent behavior
-         * to clients accessing expired values in a read-only fashion, that
-         * will say the key as non existing.
-         *
-         * Notably this covers GETs when slaves are used to scale reads. */
-        if (server.current_client &&
-            server.current_client != server.master &&
-            server.current_client->cmd &&
-            server.current_client->cmd->flags & CMD_READONLY)
-        {
-            goto keymiss;
-        }
-    }
-    val = lookupKey(db,key,flags);
-    if (val == NULL)
-        goto keymiss;
-    server.stat_keyspace_hits++;
-    return val;
-
-keymiss:
-    if (!(flags & LOOKUP_NONOTIFY)) {
-        notifyKeyspaceEvent(NOTIFY_KEY_MISS, "keymiss", key, db->id);
-    }
-    server.stat_keyspace_misses++;
-    return NULL;
+    return lookupKey(db, key, flags);
```

Comment:
If this is all `lookupKeyReadWithFlags()` does then why not chuck it and just call `lookupKey()` instead?

Reply:
This is why:

```C
/* Low level key lookup API, not actually called directly from commands
 * implementations that should instead rely on lookupKeyRead(),
 * lookupKeyWrite() and lookupKeyReadWithFlags(). */
robj *lookupKey(redisDb *db, robj *key, int flags)
```

Reply:
maybe we should prepend it with `_`, add `Internal` suffix, or remove that note about low level (i don't mind anyone calling it directly).
however, i don't want to chuck `lookupKeyRead[WithFlags]`, since i don't wanna modify all the lines that call these.
let's keep them as wrapper / convenience method.

Reply:
If we want to make it internal, we can just remove the prototype from server.h. In general, think it's good to expose as little as possible to other modules.

Reply:
thing is that we do expose the flags anyway, so i think the other wrappers are just there for either convenience or in order to avoid messing up existing lines.
i don't mind too much either way...